File: bukkit/src/main/java/xyz/lychee/dynamicdns/bukkit/BukkitMain.java
package xyz.lychee.dynamicdns.bukkit;

import dev.dejvokep.boostedyaml.YamlDocument;
import dev.dejvokep.boostedyaml.settings.dumper.DumperSettings;
import dev.dejvokep.boostedyaml.settings.general.GeneralSettings;
import dev.dejvokep.boostedyaml.settings.loader.LoaderSettings;
import lombok.Getter;
import org.bstats.bukkit.Metrics;
import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitTask;
import xyz.lychee.dynamicdns.shared.AddressUtil;
import xyz.lychee.dynamicdns.shared.DynuHook;
import xyz.lychee.dynamicdns.shared.MessageReceiver;
import xyz.lychee.dynamicdns.shared.NgrokHook;

import javax.crypto.Cipher;
import java.io.File;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class BukkitMain extends JavaPlugin {
    private static @Getter BukkitMain instance;
    private YamlDocument config;
    private BukkitTask task;
    private Metrics metrics;
    private NgrokHook ngrok;
    private DynuHook dynu;

    public void onEnable() {
        instance = this;
        this.metrics = new Metrics(this, 27000);

        try {
            this.config = YamlDocument.create(
                    new File(this.getDataFolder(), "config.yml"),
                    this.getResource("config.yml"),
                    GeneralSettings.DEFAULT,
                    LoaderSettings.builder().setCreateFileIfAbsent(true).setAutoUpdate(true).build(),
                    DumperSettings.DEFAULT
            );
            this.config.save();

            this.reload();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public void reload() throws Exception {
        this.config.reload();

        this.ngrok = new NgrokHook(this.getLogger(), new InetSocketAddress(Bukkit.getIp(), Bukkit.getPort()), this.config);
        this.dynu = new DynuHook(this.getLogger(), this.config);

        InetSocketAddress address = this.config.getBoolean("ngrok.enabled", false)
                ? AddressUtil.parseAddressFromUri(this.ngrok.load())
                : new InetSocketAddress(AddressUtil.checkPublicIp(), Bukkit.getPort());

        if (this.config.getBoolean("dynu.enabled", false)) {
            this.dynu.load(address);
        }

        this.loadSocketClient();
    }

    public void loadSocketClient() throws Exception {
        if (!this.config.getBoolean("socket_client.enabled", false)) {
            return;
        }

        String token = this.config.getString("socket_client.token");
        if (token == null) {
            this.getLogger().warning("Value socket_client.token has not been set!");
            return;
        }

        String address = this.config.getString("socket_client.target_address");
        if (address == null) {
            this.getLogger().warning("Value socket_client.target_address has not been set!");
            return;
        }

        String[] split = address.split(":");
        if (split.length != 2) {
            this.getLogger().warning("Invalid socket_client target address format. Expected format: host:port");
            return;
        }

        if (this.task != null) {
            this.task.cancel();
        }

        Cipher cipher = MessageReceiver.createCipher(token, Cipher.ENCRYPT_MODE);

        String message = this.config.getString("socket_client.name") + ":" + getServer().getPort();
        byte[] encrypted = cipher.doFinal(message.getBytes(StandardCharsets.UTF_8));

        InetSocketAddress socketAddress = new InetSocketAddress(split[0], Integer.parseInt(split[1]));

        int interval = 20 * this.config.getInt("socket_client.interval", 30);
        this.task = Bukkit.getScheduler().runTaskTimerAsynchronously(this, () -> {
            try (Socket socket = new Socket()) {
                socket.setSoTimeout(5000);
                socket.connect(socketAddress, 5000);

                socket.getOutputStream().write(encrypted);
            } catch (IOException e) {
                this.getLogger().warning("Failed to send update: " + e.getMessage());
            }
        }, interval, interval);
    }

    public void onDisable() {
        if (this.task != null) this.task.cancel();
        if (this.metrics != null) this.metrics.shutdown();
        if (this.ngrok != null) this.ngrok.unload();
    }
}

File: bukkit/src/main/kotlin/xyz/lychee/dynamicdns/bukkit/BukkitMain.kt
package xyz.lychee.dynamicdns.bukkit

import dev.dejvokep.boostedyaml.YamlDocument
import dev.dejvokep.boostedyaml.settings.dumper.DumperSettings
import dev.dejvokep.boostedyaml.settings.general.GeneralSettings
import dev.dejvokep.boostedyaml.settings.loader.LoaderSettings
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitTask
import xyz.lychee.dynamicdns.shared.AddressUtil
import xyz.lychee.dynamicdns.shared.DynuHook
import xyz.lychee.dynamicdns.shared.MessageReceiver
import xyz.lychee.dynamicdns.shared.NgrokHook
import java.io.File
import java.io.IOException
import java.net.InetSocketAddress
import java.net.Socket
import java.nio.charset.StandardCharsets
import javax.crypto.Cipher

class BukkitMain : JavaPlugin() {
    lateinit var configDoc: YamlDocument
        private set
    var task: BukkitTask? = null
    var metrics: Metrics? = null
    var ngrok: NgrokHook? = null
    var dynu: DynuHook? = null

    companion object {
        @JvmStatic
        var instance: BukkitMain? = null
            private set
    }

    override fun onEnable() {
        instance = this
        this.metrics = Metrics(this, 27000)

        try {
            this.configDoc = YamlDocument.create(
                File(this.dataFolder, "config.yml"),
                this.getResource("config.yml"),
                GeneralSettings.DEFAULT,
                LoaderSettings.builder().setCreateFileIfAbsent(true).setAutoUpdate(true).build(),
                DumperSettings.DEFAULT
            )

            this.reload()
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    @Throws(Exception::class)
    fun reload() {
        this.configDoc.reload()

        val ngrokHook = NgrokHook(this.logger, InetSocketAddress(Bukkit.getIp(), Bukkit.getPort()), this.configDoc)
        val dynuHook = DynuHook(this.logger, this.configDoc)
        this.ngrok = ngrokHook
        this.dynu = dynuHook

        val address = if (this.configDoc.getBoolean("ngrok.enabled", false)) {
            AddressUtil.parseAddressFromUri(ngrokHook.load())
        } else {
            InetSocketAddress(AddressUtil.checkPublicIp(), Bukkit.getPort())
        }

        if (this.configDoc.getBoolean("dynu.enabled", false)) {
            dynuHook.load(address)
        }

        this.loadSocketClient()
    }

    @Throws(Exception::class)
    fun loadSocketClient() {
        if (!this.configDoc.getBoolean("socket_client.enabled", false)) {
            return
        }

        val token = this.configDoc.getString("socket_client.token")
        if (token == null) {
            this.logger.warning("Value socket_client.token has not been set!")
            return
        }

        val addressStr = this.configDoc.getString("socket_client.target_address")
        if (addressStr == null) {
            this.logger.warning("Value socket_client.target_address has not been set!")
            return
        }

        val split = addressStr.split(":")
        if (split.size != 2) {
            this.logger.warning("Invalid socket_client target address format. Expected format: host:port")
            return
        }

        this.task?.cancel()

        val cipher = MessageReceiver.createCipher(token, Cipher.ENCRYPT_MODE)
        val message = "${this.configDoc.getString("socket_client.name")}:${server.port}"
        val encrypted = cipher.doFinal(message.toByteArray(StandardCharsets.UTF_8))

        val socketAddress = InetSocketAddress(split[0], split[1].toInt())

        val interval = 20L * this.configDoc.getInt("socket_client.interval", 30)
        this.task = Bukkit.getScheduler().runTaskTimerAsynchronously(this, Runnable {
            try {
                Socket().use { socket ->
                    socket.soTimeout = 5000
                    socket.connect(socketAddress, 5000)
                    socket.outputStream.write(encrypted)
                }
            } catch (e: IOException) {
                this.logger.warning("Failed to send update: ${e.message}")
            }
        }, interval, interval)
    }

    override fun onDisable() {
        this.task?.cancel()
        this.metrics?.shutdown()
        this.ngrok?.unload()
    }
}
